phylogeny/prototypeSelection.py

```diff
@@ -0,0 +1,99 @@
+consisting of exactly k elements, such that s best represents the full set S.
+
+Here, we define "best represents" as maximizing the sum of distances between
+all points, i.e. sum {over a,b in S} DM[a,b]. This is our objective function.
+
+This problem is known to be NP-hard [1], thus we need to resort to heuristics.
+In the future, this module will implement several heuristics, whose quality can
+be measured by the objective function for each problem instance, since there is
+no global winner.
+For completeness, the exact but exponential algorithm is implemented, too.
+  "prototypeSelection_exhaustive"
+
+[1] Gamez, J. Esteban, François Modave, and Olga Kosheleva.
+    "Selecting the most representative sample is NP-hard:
+     Need for expert (fuzzy) knowledge."
+    Fuzzy Systems, 2008. FUZZ-IEEE 2008.
+"""
+
+from skbio.stats.distance import DistanceMatrix
+import itertools
+import numpy as np
+import scipy as sp
+from typing import Sequence, Tuple
+
+
+def distanceSum(elements: Sequence[str], dm: DistanceMatrix) -> float:
+    '''Compute the sum of pairwise distances for the given elements according to
+    the given distance matrix.
+
+    Parameters
+    ----------
+    elements: sequence of str
+        list or elements for which the sum of distances is computed
+    dm: skbio.stats.distance.DistanceMatrix
+        pairwise distance matrix.
+
+    Returns
+    -------
+    float:
+        the sum of all pairwise distances of dm for IDs in elements
+    '''
+
+    return np.tril(dm.filter(elements).data).sum()
+
+
+def prototypeSelection_exhaustive(
+      distanceMatrix: DistanceMatrix,
+      numPrototypes: int, maxCombinationsToTest: int=200000) -> Sequence[str]:
+    '''Select k prototypes for given distance matrix
+
+    Parameters
+    ----------
+    distanceMatrix: skbio.stats.distance.DistanceMatrix
+        pairwise distances for all elements in the full set S.
+        Must be symmetric and non-hollow.
+    numPrototypes: int
+        Number of prototypes to select for distance matrix.
+        Must be >= 2, since a single prototype is useless.
+        Must be smaller than the number of elements in the distance matrix,
+        otherwise no reduction is necessary.
+    maxCombinationsToTest: int
+        The maximal number of combinations to test. If exceeding, the function
+        declines execution.
+
+    Returns
+    -------
+    sequence of str
+        A sequence holding selected prototypes, i.e. a sub-set of the
+        elements in the distance matrix.
+
+    Notes
+    -----
+    This is the reference implementation for an exact algorithm for the
+    prototype selection problem. It has an exponential runtime and will only
+    operate on small instances (< maxCombinationsToTest).
+    Idea: test all (n over k) combinations of selecting k elements from n with-
+          out replacement. Compute the objective for each such combination and
+          report the combination with maximal value.
+    '''
```

Comment:
The documentation is missing a "Raises" section, in which the raised errors are explained